File: julia/MOLE.jl/examples/elliptic/elliptic1D.jl
"""
A Julia script that solves the 1D Poisson's equation with Robin boundary
conditions using mimetic operators.

u_xx = e^x

0 < x < 1

Robin Boundary Conditions:
u(0) - u_x(0) = 0
u(1) + u_x(1) = 2 e

Exact Solution:
u(x) = e^x
"""

using Plots
import MOLE: Operators, BCs

# Domain limits
west = 0.0
east = 1.0

k = 6 # operator order of accuracy
m = 2*k + 1 # minimum number of cells to attain desired accuracy
dx = (east-west)/m # step length

path = joinpath(@__DIR__, "output") # Output path to store generated plots
mkpath(path)

L = Operators.lap(k,m,dx)

# Impose Robin boundary condition on laplacian operator
a = 1.0
b = 1.0
L = L + BCs.robinBC(k,m,dx,a,b)

# 1D staggered grid
grid = [west; (west+(dx/2)):dx:(east-(dx/2)); east]

# RHS
U = exp.(grid)
U[1] = 0
U[end] = 2*exp(1)

U = L\U

# Plot result
p = Plots.scatter(grid, U, label="Approximated", show = false)
plot!(p, grid, exp.(grid), label="Analytical", show = false)
plot!(
    p,
    xlabel="x",
    ylabel="u(x)",
    title="Poisson's equation with Robin BC",
    show = false
)
Plots.png(p, joinpath(path, "elliptic1D_output.png"))

File: examples/python/elliptic1D.py
'''
    SPDX-License-Identifier: GPL-3.0-or-later
    © 2008-2024 San Diego State University Research Foundation (SDSURF).
    See LICENSE file or https://www.gnu.org/licenses/gpl-3.0.html for details.
'''
import numpy as np
import sys
import shutil
import matplotlib.pyplot as plt

from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent)+'/src/python')
from pymole import Grid, Gradient, Divergence, Laplacian, RobinBoundaryCondition

west = 0   # Domain's limits
east = 1

k = 6      # Operator's order of accuracy
m = 2*k+1  # Minimum number of cells to attain the desired accuracy

grid = Grid.generate(west, east, shape=m+1)

L = Laplacian(grid, accuracy_order=k).matrix

terminal_width, _ = shutil.get_terminal_size()
np.set_printoptions(linewidth=terminal_width, suppress=True)

BC = RobinBoundaryCondition(grid, k, dirichlet_coefficient=1.0, neumann_coefficient=1.0)
L_BC = L + BC.matrix

x = np.r_[west, np.arange(west + grid.spacing/2, east - grid.spacing/2 + grid.spacing/2, grid.spacing), east]
U = np.exp(x)[:, None]
U[0] = 0          # West BC
U[-1] = 2*np.exp(1)  # East BC
U = np.linalg.inv(L_BC.toarray()) @ U

plt.plot(x, U, 'o', label='Approximated')
plt.plot(x, np.exp(x), label='Analytical')
plt.legend(loc='upper left')
plt.title("Poisson's equation with Robin BC")
plt.xlabel('x')
plt.ylabel('u(x)')
plt.show()